File: frontend/lib/models/invoice.dart
class SalesInvoice {
  final int id;
  final String invoiceNumber;
  final int customerId;
  final double total;
  final double discount;
  final double netTotal;
  final double paid;
  final String paymentType;
  final String? notes;
  final List<SalesItem> items;

  SalesInvoice({
    required this.id,
    required this.invoiceNumber,
    required this.customerId,
    required this.total,
    required this.discount,
    required this.netTotal,
    required this.paid,
    required this.paymentType,
    this.notes,
    required this.items,
  });

  factory SalesInvoice.fromJson(Map<String, dynamic> json) {
    return SalesInvoice(
      id: json['id'],
      invoiceNumber: json['invoice_number'],
      customerId: json['customer_id'],
      total: (json['total'] as num).toDouble(),
      discount: (json['discount'] as num).toDouble(),
      netTotal: (json['net_total'] as num).toDouble(),
      paid: (json['paid'] as num).toDouble(),
      paymentType: json['payment_type'],
      notes: json['notes'],
      items: (json['items'] as List?)
              ?.map((e) => SalesItem.fromJson(e))
              .toList() ??
          [],
    );
  }
}

class SalesItem {
  final int id;
  final int productId;
  final double quantity;
  final double unitPrice;
  final double total;

  SalesItem({
    required this.id,
    required this.productId,
    required this.quantity,
    required this.unitPrice,
    required this.total,
  });

  factory SalesItem.fromJson(Map<String, dynamic> json) {
    return SalesItem(
      id: json['id'],
      productId: json['product_id'],
      quantity: (json['quantity'] as num).toDouble(),
      unitPrice: (json['unit_price'] as num).toDouble(),
      total: (json['total'] as num).toDouble(),
    );
  }
}

class PurchaseInvoice {
  final int id;
  final String invoiceNumber;
  final int supplierId;
  final double total;
  final double paid;
  final String paymentType;
  final String? notes;
  final List<PurchaseItem> items;

  PurchaseInvoice({
    required this.id,
    required this.invoiceNumber,
    required this.supplierId,
    required this.total,
    required this.paid,
    required this.paymentType,
    this.notes,
    required this.items,
  });

  factory PurchaseInvoice.fromJson(Map<String, dynamic> json) {
    return PurchaseInvoice(
      id: json['id'],
      invoiceNumber: json['invoice_number'],
      supplierId: json['supplier_id'],
      total: (json['total'] as num).toDouble(),
      paid: (json['paid'] as num).toDouble(),
      paymentType: json['payment_type'],
      notes: json['notes'],
      items: (json['items'] as List?)
              ?.map((e) => PurchaseItem.fromJson(e))
              .toList() ??
          [],
    );
  }
}

class PurchaseItem {
  final int id;
  final int productId;
  final double quantity;
  final double unitPrice;
  final double total;

  PurchaseItem({
    required this.id,
    required this.productId,
    required this.quantity,
    required this.unitPrice,
    required this.total,
  });

  factory PurchaseItem.fromJson(Map<String, dynamic> json) {
    return PurchaseItem(
      id: json['id'],
      productId: json['product_id'],
      quantity: (json['quantity'] as num).toDouble(),
      unitPrice: (json['unit_price'] as num).toDouble(),
      total: (json['total'] as num).toDouble(),
    );
  }
}

File: app/models/invoice.py
from datetime import datetime, timezone

from sqlalchemy import DateTime, ForeignKey, Integer, Numeric, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class SalesInvoice(Base):
    __tablename__ = "sales_invoices"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    invoice_number: Mapped[str] = mapped_column(String(50), unique=True, nullable=False, index=True)
    customer_id: Mapped[int] = mapped_column(Integer, ForeignKey("customers.id"), nullable=False)
    user_id: Mapped[int] = mapped_column(Integer, ForeignKey("users.id"), nullable=False)
    total: Mapped[float] = mapped_column(Numeric(12, 2), default=0)
    discount: Mapped[float] = mapped_column(Numeric(12, 2), default=0)
    net_total: Mapped[float] = mapped_column(Numeric(12, 2), default=0)
    paid: Mapped[float] = mapped_column(Numeric(12, 2), default=0)
    payment_type: Mapped[str] = mapped_column(String(20), default="cash")
    notes: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    items: Mapped[list["SalesItem"]] = relationship(back_populates="invoice")


class SalesItem(Base):
    __tablename__ = "sales_items"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    invoice_id: Mapped[int] = mapped_column(Integer, ForeignKey("sales_invoices.id"), nullable=False)
    product_id: Mapped[int] = mapped_column(Integer, ForeignKey("products.id"), nullable=False)
    quantity: Mapped[float] = mapped_column(Numeric(12, 2), nullable=False)
    unit_price: Mapped[float] = mapped_column(Numeric(12, 2), nullable=False)
    total: Mapped[float] = mapped_column(Numeric(12, 2), nullable=False)

    invoice: Mapped["SalesInvoice"] = relationship(back_populates="items")


class PurchaseInvoice(Base):
    __tablename__ = "purchase_invoices"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    invoice_number: Mapped[str] = mapped_column(String(50), unique=True, nullable=False, index=True)
    supplier_id: Mapped[int] = mapped_column(Integer, ForeignKey("suppliers.id"), nullable=False)
    user_id: Mapped[int] = mapped_column(Integer, ForeignKey("users.id"), nullable=False)
    total: Mapped[float] = mapped_column(Numeric(12, 2), default=0)
    paid: Mapped[float] = mapped_column(Numeric(12, 2), default=0)
    payment_type: Mapped[str] = mapped_column(String(20), default="cash")
    notes: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    items: Mapped[list["PurchaseItem"]] = relationship(back_populates="invoice")


class PurchaseItem(Base):
    __tablename__ = "purchase_items"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    invoice_id: Mapped[int] = mapped_column(Integer, ForeignKey("purchase_invoices.id"), nullable=False)
    product_id: Mapped[int] = mapped_column(Integer, ForeignKey("products.id"), nullable=False)
    quantity: Mapped[float] = mapped_column(Numeric(12, 2), nullable=False)
    unit_price: Mapped[float] = mapped_column(Numeric(12, 2), nullable=False)
    total: Mapped[float] = mapped_column(Numeric(12, 2), nullable=False)

    invoice: Mapped["PurchaseInvoice"] = relationship(back_populates="items")
